busbus/web.py

```diff
@@ -0,0 +1,154 @@
+from busbus.provider import ProviderBase
+
+import cherrypy
+import collections
+import itertools
+import math
+import types
+
+
+def json_handler(*args, **kwargs):
+    value = cherrypy.serving.request._json_inner_handler(*args, **kwargs)
+    return BaseEntityJSONEncoder().encode(value).encode('utf-8')
+cherrypy.config['tools.json_out.handler'] = json_handler
+
+
+EXPAND_TYPES = {
+    'providers': ProviderBase,
+    'agencies': busbus.Agency,
+    'stops': busbus.Stop,
+    'routes': busbus.Route,
+    'arrivals': busbus.Arrival,
+}
+
+
+def unexpand_init(result, to_expand):
+    return ({attr: unexpand(value, to_expand)
+             for attr, value in dict(obj).items()}
+            for obj in result)
+
+
+def unexpand(obj, to_expand):
+    for name, cls in EXPAND_TYPES.items():
+        if isinstance(obj, cls):
+            if name not in to_expand:
+                return {'id': obj.id}
+            else:
+                return {attr: unexpand(value, to_expand)
+                        for attr, value in dict(obj).items()}
+    if isinstance(obj, dict):
+        return {attr: unexpand(value, to_expand)
+                for attr, value in obj.items()}
+    if isinstance(obj, (list, tuple, collections.Iterator)):
+        return (unexpand(value, to_expand) for value in obj)
+    return obj
+
+
+class APIError(Exception):
+
+    def __init__(self, msg, error_code=500):
+        self.msg = msg
+        self.error_code = error_code
+
+
+class Engine(busbus.Engine):
+
+    def __init__(self, *args, **kwargs):
+        # perhaps fix this to use a decorator somehow?
+        self._entity_actions = {
+            ('stops', 'find'): (self.stops_find, 'stops'),
+            ('routes', 'directions'): (self.routes_directions, 'directions'),
+        }
+        super(Engine, self).__init__(*args, **kwargs)
+
+    @cherrypy.popargs('entity', 'action')
+    @cherrypy.expose
+    @cherrypy.tools.json_out()
+    def default(self, entity=None, action=None, **kwargs):
+        if entity is None:
+            return self.help()
+
+        response = {
+            'request': {
+                'status': 'ok',
+                'entity': entity,
+                'params': kwargs,
+            }
+        }
+        to_expand = (kwargs.pop('_expand').split(',')
+                     if '_expand' in kwargs else [])
+        if to_expand:
+            response['request']['expand'] = to_expand
+        if action:
+            response['request']['action'] = action
+            if (entity, action) in self._entity_actions:
+                func, entity = self._entity_actions[(entity, action)]
+                try:
+                    response[entity] = unexpand_init(func(**kwargs), to_expand)
+                except APIError as exc:
+                    response['request']['status'] = 'error'
+                    response['error'] = exc.msg
+                    cherrypy.response.status = exc.error_code
+                return response
```

Comment:
Since the "_limit" attribute is only checked in the next block, and this returns here, then any request with actions cannot utilize the "_limit" attribute.

i.e: stops/find?latitude=38.9717&longitude=-95.2353&distance=400&_limit=1
Gives me more than one stop.